hnn_core/lfp.py

```diff
@@ -0,0 +1,231 @@
+        # depending on the electrode position along the section axis.
+        a = end - start
+        norm_a = norm(a)
+        b = electrode_pos - end
+        # projection: H = a.cos(theta) = a.dot(b) / |a|
+        H = np.dot(b, a) / norm_a  # NB can be negative
+        L = H + norm_a
+        R2 = np.dot(b, b) - H ** 2  # NB squares
+        # To avoid numerical errors when electrode is placed (anywhere) on the
+        # section axis, enforce minimal axial distance
+        R2 = max(R2, (section.diam / 2.0 + 0.1) ** 2)
+
+        if L < 0 and H < 0:  # electrode is "behind" section
+            num = np.sqrt(H ** 2 + R2) - H  # == norm(b) - H
+            denom = np.sqrt(L ** 2 + R2) - L
+        elif L > 0 and H < 0:  # electrode is "on top of" section
+            num = (np.sqrt(H ** 2 + R2) - H) * (L + np.sqrt(L ** 2 + R2))
+            denom = R2
+        else:  # electrode is "ahead of" section
+            num = np.sqrt(L ** 2 + R2) + L
+            denom = np.sqrt(H ** 2 + R2) + H  # == norm(b) + H
+
+        phi = np.log(num / denom) / norm_a
+
+    # [dis]: um; [sigma]: S / m
+    # [phi / sigma] = [1/dis] / [sigma] = 1 / [dis] x [sigma]
+    # [dis] x [sigma] = um x (S / m) = 1e-6 S
+    # transmembrane current returned by _ref_i_membrane_ is in [nA]
+    # ==> 1e-9 A x (1 / 1e-6 S) = 1e-3 V = mV
+    # ===> multiply by 1e3 to get uV
+    return 1000.0 * phi / (4.0 * np.pi * sigma)
+
+
+class _LFPElectrode:
+    """Local field potential (LFP) electrode class.
+
+    The handler is set up to maintain a vector of membrane currents at at every
+    inner segment of every section of every cell on each CVODE integration
+    step. In addition, it records a time vector of sample times. This class
+    must be instantiated and attached to the network during the building
+    process. It is used in conjunction with the calculation of extracellular
+    potentials.
+
+    Parameters
+    ----------
+    coord : tuple
+        The (x, y, z) coordinates (in um) of the LFP electrode.
+    sigma : float
+        Extracellular conductivity, in S/m, of the assumed infinite,
+        homogeneous volume conductor that the cell and electrode are in.
+    method : str
+        Approximation to use. 'psa' (default) assigns all transmembrane
+        currents to the center point (0.5) (point source approximation).
+        'lsa' treats the section as a line source, and a single multiplier is
+        calculated for each section (line source approximation).
+    cvode : instance of h.CVode
+        Multi order variable time step integration method.
+
+    Attributes
+    ----------
+    lfp_v : instance of h.Vector
+        The LFP voltage (in uV).
+
+    Notes
+    -----
+    See Table 5 in http://jn.physiology.org/content/104/6/3388.long for
+    measured values of sigma in rat cortex (note units there are mS/cm)
+    """
+
+    def __init__(self, coord, sigma=0.3, method='psa', cvode=None):
```

Comment:
a clear segregation between `Neuron` and non-Neuron components might be helpful to make this class accessible to the user. In the __init__, leave out any Neuron components, probably just the locations of the electrode. Then have a method called "record" which is called from `NetworkBuilder` which has all the Neuron components. Then you can have the `plot` method I talk about above

Reply:
This is now done (methinks)